src/scatter_buildblock/ScatterEstimation.cxx

```diff
@@ -0,0 +1,1486 @@
+    this->parser.add_key("minimum scale value",
+                         &this->min_scale_value);
+    this->parser.add_key("half filter width",
+                         &this->half_filter_width);
+    this->parser.add_key("remove interleaving",
+                         &this->remove_interleaving);
+    this->parser.add_key("run in 2d projdata",
+                         &this->run_in_2d_projdata);
+}
+
+ScatterEstimation::
+ScatterEstimation()
+{
+    this->set_defaults();
+}
+
+ScatterEstimation::
+ScatterEstimation(const std::string& parameter_filename)
+{
+    if (parameter_filename.size() == 0)
+    {
+        this->set_defaults();
+        this->ask_parameters();
+    }
+    else
+    {
+        this->set_defaults();
+        if (!this->parse(parameter_filename.c_str()))
+        {
+            error("ScatterEstimation: Error parsing input file %s. Aborting.", parameter_filename.c_str());
+        }
+    }
+}
+
+bool
+ScatterEstimation::
+post_processing()
+{
+    // Check that the crusial parts have been set.
+    info("ScatterEstimation: Loading input projection data");
+    if (this->input_projdata_filename.size() == 0)
+    {
+        warning("ScatterEstimation: No input projdata filename is given. Aborting.");
+        return true;
+    }
+
+    this->input_projdata_sptr =
+            ProjData::read_from_file(this->input_projdata_filename);
+
+    // If the reconstruction_template_sptr is null then, we need to parse it from another
+    // file. I prefer this implementation since makes smaller modular files.
+    if (this->recon_template_par_filename.size() == 0)
+    {
+        warning("ScatterEstimation: Please define a reconstruction method. Aborting.");
+        return true;
+    }
+    else
+    {
+        KeyParser local_parser;
+        local_parser.add_start_key("Reconstruction");
+        local_parser.add_stop_key("End Reconstruction");
+        local_parser.add_parsing_key("reconstruction method", &this->reconstruction_template_sptr);
+        if (!local_parser.parse(this->recon_template_par_filename.c_str()))
+        {
+            warning(boost::format("ScatterEstimation: Error parsing reconstruction parameters file %1%. Aborting.")
+                    %this->recon_template_par_filename);
+            return true;
+        }
+    }
+
+    info("ScatterEstimation: Loading attenuation image...");
+    if (this->atten_image_filename.size() == 0)
+    {
+        warning("ScatterEstimation: Please define an attenuation image. Aborting.");
+        return true;
+    }
+    else
+        this->atten_image_sptr =
+            read_from_file<DiscretisedDensity<3,float> >(this->atten_image_filename);
+
+    if(is_null_ptr(multiplicative_binnorm_sptr))
+        warning("ScatterEstimation: No multiplicative coefficients have been set!!\n\
+                At least attenuation has to be set!");
+
+    this->multiplicative_binnorm_sptr->set_up(this->input_projdata_sptr->get_proj_data_info_sptr()->create_shared_clone());
```

Comment:
this `set_up` has to be moved to `ScatterEstimation::set_up`. In the future, you don't know if things will be parsed, therefore if `post_processing` will be called.

Reply:
ok